crates/bevy_tasks/src/bevy_executor.rs

```diff
@@ -0,0 +1,1296 @@
+
+    // The thread local state of the executor for the current thread.
+    local_state: &'a ThreadLocalState,
+}
+
+impl Runner<'_> {
+    /// Creates a runner and registers it in the executor state.
+    fn new(state: &State) -> Runner<'_> {
+        let local_state = THREAD_LOCAL_STATE.get_or_default();
+        let runner = Runner {
+            state,
+            ticker: Ticker::new(state),
+            ticks: 0,
+            local_state,
+        };
+        state
+            .stealer_queues
+            .write()
+            .unwrap_or_else(PoisonError::into_inner)
+            .push(&local_state.stealable_queue);
+        runner
+    }
+
+    /// Waits for the next runnable task to run.
+    fn runnable(&mut self, _rng: &mut fastrand::Rng) -> impl Future<Output = Runnable> {
+        // SAFETY: The provided search function does not access LOCAL_QUEUE in any way, and thus cannot 
+        // alias.
+        let runnable = unsafe {
+            self
+            .ticker
+            .runnable_with(|tls| {
+                if let Some(r) = tls.local_queue.pop_back() {
+                    return Some(r);
+                }
+
+                crate::cfg::multi_threaded! {
+                    if {
+                        // Try the local queue.
+                        if let Ok(r) = self.local_state.stealable_queue.pop() {
+                            return Some(r);
+                        }
+
+                        // Try stealing from the global queue.
+                        if let Ok(r) = self.state.queue.pop() {
+                            steal(&self.state.queue, &self.local_state.stealable_queue);
+                            return Some(r);
+                        }
+
+                        // Try stealing from other runners.
+                        if let Ok(stealer_queues) = self.state.stealer_queues.try_read() {
+                            // Pick a random starting point in the iterator list and rotate the list.
+                            let n = stealer_queues.len();
+                            let start = _rng.usize(..n);
+                            let iter = stealer_queues
+                                .iter()
+                                .chain(stealer_queues.iter())
+                                .skip(start)
+                                .take(n);
+
+                            // Remove this runner's local queue.
+                            let iter =
+                                iter.filter(|local| !core::ptr::eq(**local, &self.local_state.stealable_queue));
+
+                            // Try stealing from each local queue in the list.
+                            for local in iter {
+                                steal(*local, &self.local_state.stealable_queue);
+                                if let Ok(r) = self.local_state.stealable_queue.pop() {
+                                    return Some(r);
+                                }
+                            }
```

Comment:
Nit, because this is from the original implementation, but I somewhat disagree with the ordering here. I'm pretty sure rayon takes from the local queue, then from peers, then from the global queue, whereas this is local-global-peers. The rational for this is that it's better to finish your current work (stuff already queued on workers) before you start something new (by pulling from the injector queue).

Can we run some benchmarks with alternative orderings at some point?